Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/DccValidationRepository.kt

```diff
@@ -1,33 +1,95 @@
 package de.rki.coronawarnapp.covidcertificate.validation.core
 
+import com.google.gson.Gson
 import de.rki.coronawarnapp.covidcertificate.validation.core.country.DccCountry
-import de.rki.coronawarnapp.covidcertificate.validation.core.country.DccCountryRepository
-import de.rki.coronawarnapp.covidcertificate.validation.core.rule.DccValidationRuleRepository
+import de.rki.coronawarnapp.covidcertificate.validation.core.server.DccValidationServer
+import de.rki.coronawarnapp.covidcertificate.validation.core.rule.DccValidationRule
+import de.rki.coronawarnapp.covidcertificate.validation.core.rule.server.DccValidationRulesServer
+import de.rki.coronawarnapp.util.coroutine.AppScope
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.flow.HotDataFlow
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.util.serialization.fromJson
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.plus
+import org.joda.time.Duration
+import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
+/**
+ * Returns the latest set of dcc-country data and rules.
+ * If the user is within the validation flow, we want the cached version.
+ * Before the user enters the validation flow, refresh needs to be called and only proceeded if successful
+ *
+ * General validation flow can happen offline,
+ * but we want to make sure that we have the latest data set before starting.
+ * So call [refresh], wait for the result before proceeding and handle errors, then work with the cache [dccCountries].
+ */
 @Singleton
 class DccValidationRepository @Inject constructor(
-    private val dccCountryRepository: DccCountryRepository,
-    private val validationRulesRepository: DccValidationRuleRepository,
+    @AppScope private val appScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+    @BaseGson private val gson: Gson,
+    private val countryServer: DccValidationServer,
+    private val localCache: DccValidationCache,
+    private val rulesServer: DccValidationRulesServer,
 ) {
 
-    val dccCountries: Flow<List<DccCountry>>
-        get() = dccCountryRepository.dccCountries
+    private val internalCountries: HotDataFlow<List<DccCountry>> = HotDataFlow(
```

Comment:
Would it make sense to create a common container for both countries and rules?
Do we always need the latest version of both?

```kotlin
private val internalCountries: HotDataFlow<DccValidationData> 
```

We could still keep

```
val dccCountries: Flow<List<DccCountry>> = internalCountries.data.map { it.countries }
```
Just thinking out loud, what do you think? :thinking:

Reply:
i was thinking the same thing

Reply:
introduced new general data container